packages/common/src/plugins.ts

```diff
@@ -0,0 +1,35 @@
+import { Plugin as EnvelopPlugin, PromiseOrValue } from '@envelop/core'
+import { GraphQLParams } from './types'
+
+export type Plugin<
+  PluginContext extends Record<string, any> = {},
+  TServerContext = {},
+> = EnvelopPlugin<PluginContext> & {
+  onRequestParse?: OnRequestParseHook<TServerContext>
+}
+
+export type OnRequestParseHook<TServerContext> = (
+  payload: OnRequestParseEventPayload<TServerContext>,
+) => PromiseOrValue<void | OnRequestParseHookResult>
+
+export type RequestParser = (request: Request) => PromiseOrValue<GraphQLParams>
+
+export interface OnRequestParseEventPayload<TServerContext> {
+  serverContext: TServerContext | undefined
+  request: Request
+  requestParser: RequestParser
+  setRequestParser: (parser: RequestParser) => void
```

Comment:
My initial thought here was that you "register" a request parser here instead of setting it.

```
type RequestParser = {
  is: (request: Request) => boolean
  parse: (request: Request) => Promise<GraphQLRequestPayload>
}
```

And then the Yoga Code iterates through all the registered request parsers and calls .is until there is a match.

Reply:
The iteration was a problem in the previous implementation. It stops the iteration if `is` returns true. For example, we have to put multipart parser before the regular POST json parser because POST parser's `is` returns true without checking content-type. And this makes it impossible to override that unless you put your parser before another.
I think current approach is similar what we have in Envelop. Plugins are called sequentially and each following one can replace the previous ones' behavior.

Reply:
Okay how about the following:

```
function useRequestParser(requestParser: RequestParser): Plugin {
    return {
    onRequestParse({ request, setRequestParser }) {
      if (requestParser.is(request)) {
        setRequestParser(requestParser)
      }
    },
  }
}

Reply:
That way the packages RequestParser can still be used as standalone without yoga and this plugin can be used for adding them.

Reply:
I agree with your concerns!

Reply:
In this case, `RequestParser` signature would be weird because it has `is` method that is called only with this plugin while the internals are calling its `parse` method without caring `is`.
I'm fine with separating the actual logic and the Yoga specific plugin but not sure it is going to look like.

Reply:
oh sorry, I actually mean everything stays as you did so far, I agree with your concerns. Only the logic should not be hard-coded within the plugin